### PR TITLE
feat(contracts): lazy merkle tree, reusable merkle tree tests

### LINKED ISSUE #389: Add tests for `ReusableMerkleTree`

### LINKED ISSUE #411: Lazily build merkle tree roots onchain
Right now when we add an item to a merkle tree we calculate the root immediately. This means every addition takes `HEIGHT` hashes, regardless of how full the tree is.

Instead we can build the smallest possible subtree upon insertion. For example, when we insert the first element we simply do `H(element, 0)` to calculate the first parent. If we want to access the root we'll then hash up the height of the tree to get the root.

As we insert elements the complexity increases until the tree is almost full, at which point we have the same complexity as a normal implementation. The amortized cost should be ~25% less than a normal implementation. Early insertions/updates in the tree are _much_ cheaper, with later insertions costing the same as a normal implementation.

This works well for trees that are changed often, but rarely have their root accessed. This is good for the epoch tree in attesters. This tree is changed often, but the root only needs to be accessed after an epoch ends, at which point the tree is immutable (so the final root can be calculated once and stored).

A visual of adding elements to a tree of depth 3

Insertion cost: 0 hashes
```
a
```

Insertion cost: 1 hash
```
  r1
 /  \
a    b
```

Insertion cost: 2 hashes
```
      r3
   /       \
  s0       s1
 /  \    /    \
a   b   c      0
```

Insertion cost: 2 hashes
```
      r4
   /       \
  s0       s1
 /  \    /    \
a   b   c      d
```

Insertion cost: 3 hashes (s3 is a constant here)
```
              r5
        /              \
      s4                s5
   /       \         /        \
  s0       s1       s2        s3
 /  \    /    \    /   \    /    \
a   b  c       d   e   0    0      0
```

Insertion cost: 3 hashes (s3 is a constant here)
```
              r6
        /              \
      s4                s5
   /       \         /        \
  s0       s1       s2        s3
 /  \    /    \    /   \    /    \
a   b  c       d   e   f    0      0
```

Insertion cost: 3 hashes (s3 is no longer constant, but s2 is)
```
              r6
        /              \
      s4                s5
   /       \         /        \
  s0       s1       s2        s3
 /  \    /    \    /   \    /    \
a   b  c       d   e   f    g      0
```

For a tree of depth 3 a normal implementation needs 24 hashes to fill the tree. Using the lazy approach we only need 17 hashes.

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #411 
Closes #389 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by vimwitch** [comment]
The insertion cost for the lazy merkle tree is nearly constant with depth. The cost of calculating the root is (in the worst case) the same as inserting into a normal tree. The lazy merkle tree is also flexible depth, a root can be calculated for any depth, so as the tree expands the depth can logically be increased without changing the data structure.

For trees that are append only proofs can be made against subtrees to avoid the cost/race condition of calculating the latest root.

### Benchmarks

Filling a tree of depth 5
```
Average insertion cost (lazy): 77,096
Average insertion cost (normal): 230,921
```

Filling a tree of depth 9
```
Average insertion cost (lazy): 84,668
Average insertion cost (normal): 387,346
```

Filling a tree of depth 10
```
Average cost (lazy): 85,080
Average cost (normal): 428,364
```

Inserting 128 leaves into a tree of depth 32
```
Average insertion cost (lazy): 85,185
Average insertion cost (normal): 1,386,428
```

**Comment by vimwitch** [comment]
Actually the insertion cost is constant, the amortized cost for a tree just changes as the number of leaves inserted changes.

File: packages/contracts/contracts/libraries/LazyMerkleTree.sol [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            // it's a left element so we don't hash until there's a right element
+            if (index & 1 == 0) break;
+            uint40 elementIndex = indexForElement(i, index - 1);
+            hash = PoseidonT3.hash([self.elements[elementIndex], hash]);
+            unchecked {
+                index >>= 1;
+                i++;
+            }
+        }
+    }
```
Comment (vimwitch) [inline packages/contracts/contracts/libraries/LazyMerkleTree.sol:58]: Need to add an update function

File: packages/contracts/contracts/libraries/LazyMerkleTree.sol [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+        for (uint8 i = 0; i < depth; ) {
+            if (index & 1 == 0) {
+                levels[i + 1] = PoseidonT3.hash([levels[i], defaultZero(i)]);
+            } else {
+                // if the element is a right sibling the parent is guaranteed to be calculated
+                uint256 parent = self.elements[
+                    indexForElement(i + 1, index >> 1)
+                ];
+                if (parent != 0) {
```
Comment (vimwitch) [inline packages/contracts/contracts/libraries/LazyMerkleTree.sol:83]: Don't rely on a zero check for the parent if the tree is reusable. Instead track how many elements have been inserted on each level.

**Comment by vimwitch** [comment]
Updating a lazy merkle tree is 33% cheaper in the _worst_ case. If the tree is less full it's even cheaper.

Updating a tree of depth 6 repeatedly as it's filled
```
Average update cost (lazy): 155,399
Average update cost (normal): 240,408
```

These are benchmarks for attestations using the same configuration (epoch tree depth = 17).

Normal tree
```bash
·············|··················|·············|·············|····················|·············|················
|  Contract  ·  Method          ·  Min        ·  Max        ·  Avg               ·  # calls    ·  eur (avg)    │
·············|··················|·············|·············|····················|·············|················
|  Unirep    ·  attest          ·     349662  ·    1237754  ·            511204  ·         37  ·            -  │
·············|··················|·············|·············|····················|·············|···············
```

Lazy tree
```bash
·············|··················|·············|·············|····················|·············|················
|  Contract  ·  Method          ·  Min        ·  Max        ·  Avg               ·  # calls    ·  eur (avg)    │
·············|··················|·············|·············|····················|·············|················
|  Unirep    ·  attest          ·     255880  ·     418455  ·            320846  ·         33  ·            -  │
·············|··················|·············|·············|····················|·············|···············
```

**Review by vivianjeng — APPROVED** [review]
LGTM
Thank you 🙏🏻

**Pull request merged**